staging/src/k8s.io/apiserver/pkg/cel/library/library_compatibility_test.go

```diff
@@ -66,3 +65,40 @@ func TestLibraryCompatibility(t *testing.T) {
 		t.Errorf("Expected all functions in the libraries to be assigned to a kubernetes release, but found the missing function names: %v", missing)
 	}
 }
+
+// TestTypeRegistration ensures that all custom types defined and used by Kubernetes CEL libraries
+// are returned by library.Types().  Other tests depend on Types() to provide an up-to-date list of
+// types declared in a library.
+func TestTypeRegistration(t *testing.T) {
+	for _, lib := range KnownLibraries() {
+		registeredTypes := sets.New[*cel.Type]()
+		usedTypes := sets.New[*cel.Type]()
+		// scan all registered function declarations for the library
+		for _, fn := range lib.declarations() {
+			fn, err := decls.NewFunction("placeholder-not-used", fn...)
+			if err != nil {
+				t.Fatal(err)
+			}
+			for _, o := range fn.OverloadDecls() {
+				// ArgTypes include both the receiver type (if present) and
+				// all function argument types.
+				for _, at := range o.ArgTypes() {
+					switch at.Kind() {
+					// User defined types are either Opaque or Struct.
+					case types.OpaqueKind, types.StructKind:
+						usedTypes.Insert(at)
+					default:
+						// skip
+					}
+				}
+			}
+		}
+		for _, t := range lib.Types() {
```

Comment:
should we check all types start with `kubernetes.` (except possibly allowlisted exceptions like net.IP we're not sure about changing yet?)

Reply:
Nice yes. I've added library and type name checks to the tests and put `net.IP` and `net.CIDR` in an exception list.  Tristan was interested in seeing those two types promoted upstream, which is why we left the `kubernetes` package prefix off of them. I'll follow up on that.